### ISSUE TITLE
Why do we need to stop gradient for level_boxes and box_indices in PyramidROIAlign?

### ISSUE THREAD

**Issue by Superlee506** [matterport/Mask_RCNN#343]
![image](https://user-images.githubusercontent.com/13114675/37558127-97131c5e-29e5-11e8-96e0-5fa8ad81209b.png)
As in [Line 384-385](https://github.com/matterport/Mask_RCNN/blob/master/model.py#L384-L385)， why we need the stop operation?

**Comment by spring527** [comment]
index

**Comment by spring527** [comment]
index

**Comment by spring527** [comment]
index

**Comment by waleedka** [comment]
If we don't stop the gradients, TensorFlow will try to compute the gradients all the way back to the code that generates the anchor box refinement. But we already handle learning the anchor refinement in the RPN, so we don't want to influence that with additional gradients from stage 2. So, the sooner we stop it, the more unnecessary computation we avoid. Further, it's not clear (at least I haven't looked into it) how the gradients calculation back through crop_and_resize works.

**Comment by Superlee506** [comment]
@waleedka Thanks for your reply, but if we stop the gradients, how are the fpn parameters be trained. As I know the fpn is behind the ROI layer.

**Comment by zhang-qiang-github** [comment]
> I think the following phrase from tensorflow documentation about tf.stop_gradient helps:
> 
> > This is useful any time you want to compute a value with TensorFlow but need to pretend that the value was a constant.
> 
> Simply speaking, you can just recognize level_boxes and box_indices as constants. The rest are trained and updated as usual.

@JingyuQian In this way, should we stop gradient for many variable? For example, the positive_indices in the detection_targets_graph?

**Comment by zhang-qiang-github** [comment]
@waleedka Then, why should we stop gradient for the box_indices? Would box_indices affect the gradient?

**Comment by JingyuQian** [comment]
@zhang-qiang-github stopping propagation to ```box_indices``` is probably because that is handled by RPN. In a typical training scheme, you will want to freeze RPN to train various heads(regression, classification, etc.). So stopping gradient should help preventing RPN from changing.
(At least this is what I can understand. Or maybe I'm wrong. I haven't used this repo for a long time)
